rfd/0156-notifications.md

```diff
@@ -0,0 +1,404 @@
+
+## Details
+
+Notifications will be stored in the backend database and be of two types:
+
+1. User-specific notifications: Targets a single specific user. An example is a notification to notify a user that their access request has been approved.
+   These notifications will be stored in the database under the key `/notifications/user/<the user's username>/<notification id>`.
+2. Global notifications: Targets multiple users. An example is a notification to notify users that there is a new access request needing review.
+   These notifications will be stored in the database under the key `/notifications/global/<notification id>`.
+   They can be configured to target users based on one of the following matchers:
+   1. Permissions: Targets all users with the specified RBAC permission(s).
+   2. Role: Targets all users with the specified role(s).
+   3. All: Targets all users.
+
+Since user-specific notifications are stored by username, notifications for SSO users will persist between sessions.
+
+Notifications will have a default TTL of 30 days, after which they will be automatically deleted. In some cases, notifications may also automatically be deleted prior to their TTL, for
+example, a notification reminding a user to complete their periodic access list review will be deleted once the user does it.
+
+#### Types and Protos
+
+###### Notification Item
+
+These will be stored under `/notification/user/<username>/<notification id>`. They can also also be nested in a `GlobalNotification`.
+
+```protobuf
+// Notification represents a notification item.
+message Notification {
+   // ID is the unique identifier for the notification.
+   string ID = 1 [(gogoproto.jsontag) = "id"];
+   // Code is the unique notification code.
+   string Code = 2 [(gogoproto.jsontag) = "code"];
+   // Title is the title/header of the notification.
+   string Title = 3 [(gogoproto.jsontag) = "title"];
+   // Description is the text content of the notification.
+   string Description = 4 [(gogoproto.jsontag) = "description"];
+   // Created is when the notification was created, in UNIX time.
+   google.protobuf.Timestamp Created = 5 [(gogoproto.jsontag) = "created"];
+   // TTL is how long the notification should last for before automatically expiring. Default is 30 days.
+   int64 TTL = 6 [(gogoproto.jsontag) = "ttl,omitempty"];
+   // Metadata is optional additional metadata to be added to the notification.
+   google.protobuf.Struct Metadata = 7 [(gogoproto.jsontag) = "metadata,omitempty"];
+}
+```
+
+###### User Notification Preferences Item
+
+Each user will have one `UserNotificationStates` object which will keep track of the notifications they have seen or dismissed.
+
+These will be stored once per user in the database under `/notifications/states/<username>`
+
+```protobuf
+// UserNotificationStates represents a user's notification states. This is to keep track
+// of the notifications the user has seen and dismissed.
+message UserNotificationStates {
```

Comment:
> A single backend item shouldn't really be bigger than 128KiB or so

Is this a hard cap that would break things if surpassed, or just a recommended max?

> some sort of max on the number of notifications per-user, and then logic to allow newer notifications to replace older notifications if we're at capacity

This could introduce a UX issue since replacing older notifications in the list would effectively be marking them as unread/undismissing them. This would mean that users would once again start seeing notifications which they had already marked as read or dismissed. I think a potential approach that would prevent this is to have multiple "pages" of a user's states, we would store them in `notifications/states/<user>/1`, and if they reach the max we would create a `/notifications/states/<user>/2` entry to store additional ones, and repeat this as many times as needed to store all of them.